logstash-core/src/main/java/org/logstash/ext/JrubyAckedWriteClientExt.java

```diff
@@ -41,66 +43,51 @@ public final class JrubyAckedWriteClientExt extends JRubyAbstractQueueWriteClien
 
     private JRubyAckedQueueExt queue;
 
-    private AtomicBoolean closed = new AtomicBoolean();
-
     @JRubyMethod(meta = true, required = 2)
     public static JrubyAckedWriteClientExt create(final ThreadContext context, final IRubyObject recv,
         final IRubyObject queue, final IRubyObject closed) {
         return new JrubyAckedWriteClientExt(
             context.runtime, RubyUtil.ACKED_WRITE_CLIENT_CLASS,
-            queue.toJava(
-                JRubyAckedQueueExt.class
-            ),
-            closed.toJava(AtomicBoolean.class)
+            queue.toJava(JRubyAckedQueueExt.class)
         );
     }
 
-    public static JrubyAckedWriteClientExt create(final JRubyAckedQueueExt queue, final AtomicBoolean closed) {
-        return new JrubyAckedWriteClientExt(
-            RubyUtil.RUBY, RubyUtil.ACKED_WRITE_CLIENT_CLASS, queue, closed);
+    public static JrubyAckedWriteClientExt create(final JRubyAckedQueueExt queue) {
+        return new JrubyAckedWriteClientExt(RubyUtil.RUBY, RubyUtil.ACKED_WRITE_CLIENT_CLASS, queue);
     }
 
     public JrubyAckedWriteClientExt(final Ruby runtime, final RubyClass metaClass) {
         super(runtime, metaClass);
     }
 
     private JrubyAckedWriteClientExt(final Ruby runtime, final RubyClass metaClass,
-        final JRubyAckedQueueExt queue, final AtomicBoolean closed) {
+        final JRubyAckedQueueExt queue) {
         super(runtime, metaClass);
         this.queue = queue;
-        this.closed = closed;
     }
 
     @Override
     protected JRubyAbstractQueueWriteClientExt doPush(final ThreadContext context,
         final JrubyEventExtLibrary.RubyEvent event) {
-        ensureOpen();
         queue.rubyWrite(context, event.getEvent());
         return this;
     }
 
     @Override
     protected JRubyAbstractQueueWriteClientExt doPushBatch(final ThreadContext context,
         final Collection<JrubyEventExtLibrary.RubyEvent> batch) {
-        ensureOpen();
         for (final IRubyObject event : batch) {
             queue.rubyWrite(context, ((JrubyEventExtLibrary.RubyEvent) event).getEvent());
         }
         return this;
     }
 
-    private void ensureOpen() {
-        if (closed.get()) {
-            throw new IllegalStateException("Tried to write to a closed queue.");
-        }
-    }
-
     @Override
     public void push(Map<String, Object> event) {
         try {
             queue.write(new Event(event));
         } catch (IOException e) {
-            throw new IllegalStateException(e);
+            throw new QueueRuntimeException(QueueExceptionMessages.WRITE_TO_CLOSED_QUEUE, e);
```

Comment:
Yeah agree. More than that, my main concern was, except the `IOException`, we could get various exception types. For example:
- `Page#write()->write() -> IllegalStateException` (`this.headPage`)
- `Condition.signal() -> IllegalMonitorStateException` (`notEmpty`)

This is why I have wrapped exception to be catchable in the pipeline bus. One thing I am seeing as an improvement, I could give a better descriptive message instead of closed queue, say `exception while writing to a queue` and original _any_ type of exception wrapped by `QueueRuntimeException(QueueExceptionMessages.WRITE_TO_CLOSED_QUEUE, e)` would provide extra details.

Any better option are you seeing here?

Reply:
🤦 Yeah agreed, it doesn't make sense to retry the process (through `input.rb`) on `IOException` cases.